### DISCUSSION TITLE
Is it possible to filter data from api?

### DISCUSSION THREAD

**Discussion by RajathShanmugaiah** [react-page/react-page#1037] [Q&A] [answered]
I have the data coming in from the API so is it posiible to filter them in the autoform or any other suggestion?

for eg: I have a lot of products I want to choose the category and show it in the component using the autoform or any other suggestion by you

**Comment by macrozone** [comment, marked as answer]
@RajathShanmugaiah can you give more information on what you want to achieve and what you already did? 

I guess you want to have something like a ProductList that loads products from some API and then you want content-editors to be able to add this ProductList to the editor. The content-editors should be able to select a category to show only matching products.

Basically you should have this component (that has nothing to do with react-page):

```
const ProductList: <React.FC<{category: string}> = ({category}) => {
   // loads some products by a category from an api
   const products = useProducts({category});

  return <div>....

}
```

Then you can create a CellPlugin:

```

const productListPlugin: CellPlugin<{
  category: string;
}> = {
  Renderer: props => <ProductList category={props.data?.category} />,
  id: "product-list",
  version: 1,
  title: "Product List",
  description: "A product List",
  controls: {
    type: "autoform",
    schema: {
      required: ["category"],
      properties: {
        category: {
          type: "string",
          enum: ["category-1", "category-2", "category-3"] // if you have a fixed list of categories
        },
      },
    },
  },
};
```

Or if the ist of categories also come from the api, create a custom form field and use that:

```
// in the plugin definition
 schema: {
      required: ["category"],
      properties: {
        category: {
          type: "string",
          uniforms: {
             component: CategorySelectField
          }
        },
      },
    },
```

Wheras CategorySelectField is a custom uniforms-field. See docs for reference: https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-field-components-example

**Reply by RajathShanmugaiah** [reply]
@macrozone 
```
// initially I load the data using context and wrap editor with it
   <UserProvider value={apiMenuData}>
      <Editor languages={LANGUAGES} cellPlugins={cellPlugins} value={value} onChange={setValue}/>
      <Button onClick={reset}>Reset</Button>
    </UserProvider>
```
Later in the custom cell plugin I use this schema to render my custom selector 
```// in the plugin definition
 schema: {
      required: ["category"],
      properties: {
        category: {
          type: "string",
          uniforms: {
             component: CategorySelectField
          }
        },
      },
    },
```
```//here is my custom form component which is used to select the dynamic categories
const CategorySelectField= (props:any) => {
    const apiMenu = useContext(CustomContext)
    const [state, setState] = React.useState<any>(menuTypes);

const handleChange = (event: any, x: any[]) => {
    setState((newState: any[] ) => (
        newState.includes(x)
            ? newState.filter(c => c !== x)
            : [...newState, x]
        ));
  };
  menuTypes = state
    return(
        <>
         <FormGroup row>
            {
                apiMenu.types.map((type: any,index: React.Key) =>(
                    <FormControlLabel
                        key={index}
                        control={<Checkbox checked={state.includes(type)} onChange={event => handleChange(event, type)}  name={type} />}
                        label={type}
                    />
                ))
            }
            </FormGroup>
    </>)
}```
So basically I have done the steps as you have mentioned earlier but for some reason the data is not getting filtered on selection  of the dynamic categories rendered in the form 
Im also not able to hand down the types inside the renderer

**Reply by macrozone** [reply]
if you go to  https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-field-components-example and scroll down, you see an example of an `ImageUploadField`

As you see, you have to use `connectField` from uniforms to create a form field. `connectField^ injects `value` and `onChange` into your component, so you usually do not need `useState` inside of your component.

Just make sure that `value` and the argument you pass to `onChange` has the same type as required by the schema.

So in your case its a string.

Best check the uniforms docs if you have further questions about it: https://uniforms.tools/
